### ISSUE TITLE
Support conditional coroutine check with any/all

### ISSUE THREAD

**Issue by JacobHayes** [microsoft/pyright#9129]
**Is your feature request related to a problem? Please describe.**
I was writing some code that checked that `all` or `any` calls to a collection of async functions returned True. Naively, I first wrote this as (or maybe had this before refactoring to async?):

```python
all(some_async_func(x) for x in data)
```

This type checks just fine, however at runtime it is *always truthy* (and raise the "unwaited coroutine" warning) as the coroutines are never awaited and instead the coroutine itself is considered truthy.

An alternate proper form is instead:
```python
all([await some_async_func(x) for x in data])
```

**Describe the solution you’d like**
I'd like calls to `any`/`all` with coroutines to be flagged with `reportUnnecessaryComparison` ("Conditional expression references coroutine which always evaluates to True"), which _is_ flagged in the simpler case of:

```python
from typing import Any

async def check(x: Any):
    return bool(x)

if check(True): # flags reportUnnecessaryComparison
    print("yup")
```

I'm assuming this would require some special casing for `any` and `all` since functions in the general case might otherwise handle coroutines. Aside from that, I'm guessing it'd need to handle them being shadowed by a local name.

**Comment by erictraut** [comment]
From the perspective of a type checker, `any` and `all` are just function calls. It's hard to justify treating them differently from other functions in a type checker. This is definitely not a type violation. This may be something that a linter rule might cover, but it's not something that we'd be likely to add to pyright.

**Comment by JacobHayes** [comment]
Makes sense that the special casing isn't desired, but it seems like `if check(True):` isn't a type violation either. Does that make the cost/benefit cut to warrant an error from a frequency perspective? Want to get a sense of when to open issues vs consider them too niche. :)
